aten/src/ATen/native/cuda/BatchLinearAlgebra.cu

```diff
@@ -0,0 +1,372 @@
+    n, info_array, batch_size, inverse_magma_queue);
+
+  for (int64_t i = 0; i < batch_size; i++) {
+    infos[i] = info_array[i];
+  }
+#endif
+}
+
+// Because this is out-of-place inverse, the predefined macros will
+// not work
+Tensor _inverse_helper_cuda(const Tensor& self) {
+  std::vector<int64_t> infos(batchCount(self), 0);
+  auto self_working_copy = cloneBatchedColumnMajor(self);
+  auto self_inv_working_copy = cloneBatchedColumnMajor(self);
+  AT_DISPATCH_FLOATING_TYPES(self.type(), "inverse", [&]{
+    apply_inverse<scalar_t>(
+      self_working_copy, self_inv_working_copy, infos);
+  });
+  batchCheckErrors(infos, "inverse");
+  return self_inv_working_copy;
+}
+
+template <typename scalar_t>
+static void apply_potrf(Tensor& A, bool upper) {
+#ifndef USE_MAGMA
+AT_ERROR("potrf: MAGMA library not found in "
+    "compilation. Please rebuild with MAGMA.");
+#else
+  auto A_data = A.data<scalar_t>();
+  auto A_mat_stride = matrixStride(A);
+
+  magma_int_t batch_size = magma_int_cast(batchCount(A), "batchCount");
+  magma_int_t n = magma_int_cast(A.size(-2), "A.size(-2)");
+
+  magma_int_t* info_array;
+  scalar_t** A_array;
+
+  ALLOCATE_ARRAY(info_array, magma_int_t, batch_size, A);
+  ALLOCATE_ARRAY(A_array, scalar_t*, batch_size, A);
+
+  // Set up the created arrays
+  for (int64_t i = 0; i < batch_size; i++) {
+    A_array[i] = &A_data[i * A_mat_stride];
+  }
+
+  magmaPotrfBatched<scalar_t>(upper ? MagmaUpper : MagmaLower, n, A_array, n,
+			      info_array, batch_size, createMagmaQueue(A));
+
+  for (int64_t i = 0; i < batch_size; i++) {
+    AT_CHECK(info_array[i] == 0, "potrf: the leading minor of order ", info_array[i], " of argument ", i, " is not positive definite");
+  }
+#endif
+}
+
+Tensor potrf_cuda(const Tensor& self, bool upper) {
+  Tensor result;
+  if (upper) { // work around magma not doing upper
+    result = cloneBatchedColumnMajor(self.transpose(-1, -2));
+  } else {
+    result = cloneBatchedColumnMajor(self);
+  }
+  AT_DISPATCH_FLOATING_TYPES(result.type(), "potrf", [&]{
+      apply_potrf<scalar_t>( result, false);
+  });
+  if (upper) {
+    result.transpose_(-1, -2).triu_();
+  } else {
+    result.tril_();
+  }
+  return result;
+}
+
+
+template <typename T, bool upper>
+struct BatchTensorTriOp {
+  BatchTensorTriOp(T *start_, int64_t stride_batch_, int64_t stride_row_, int64_t stride_col_, int64_t k_)
+    : start(start_), stride_row(stride_row_), stride_col(stride_col_), k(k_), stride_batch(stride_batch_),
+      stride_max(stride_row_ > stride_col_? stride_row_ : stride_col_), stride_min(stride_row_ < stride_col_? stride_row_ : stride_col_) {}
+
+  __device__ __forceinline__ int mask(T *result) {
+    ptrdiff_t n = result - start;
+    if (stride_batch > stride_max) {
+      n = n % stride_batch;
+    } else if (stride_batch > stride_min) {
+      n = n - n % stride_max + n % stride_batch; // eliminate batch part
+    } // if stride_batch < stride min, the divisions below will eliminate batch
```

Comment:
@t-vi Could you please give a small explanation of the logic here? It doesn't seem very obvious to me, unfortunately. Thank you.

Reply:
Here, you have `stride_max > stride_batch > stride_min` you now want to eliminate the batch contribution in the offset. Subtracting `n % stride_max` will eliminate both the batch and the "lower" contribution, so adding `n % stride_batch` adds the second part back. As a result exactly the batch offset is removed.

Reply:
I got it. Thank you very much.